proposals/0018/README.md

```diff
@@ -0,0 +1,191 @@
+    uint32 cidr = 2;
+    string gateway = 3;
+
+    enum Family {
+      UNKNOWN = 0;
+      V4 = 1;
+      V6 = 2;
+    }
+    Family family = 4;
+
+    // fork in the road! choose a path
+    // path 2Left
+    enum Flags {
+      NONE = 0;
+      PUBLIC = 1;
+      MANAGEMENT = 2;
+    };
+    uint32 flags = 5;  // bitwise-or of Flags
+
+    // vs
+    //path 2Right
+    bool public = 5;
+    bool management = 6;
+
+  }
+  repeated IP ips = 6;
+
+  string userdata = 7;
+  string crypted_root_password = 8;
+
+  repeated string tags = 9;
+  repeated string ssh_keys = 10;
+}
+```
+
+### Forks in the Roads
+
+#### Fork 1: OSes
+
+Custom iPXE is not just another OS.
+It has a very fundamental difference from anything else we might run with Workflows, the iPXE script (or its URL) is an input parameter.
+Any other workflow would always want to boot into OSIE in workflow mode or otherwise be known statically in the Workflow Template.
+But we just don't have a need to specify an iPXE script as part of the input for non-CustomIPXE OSes.
+
+EquinixMetal supports Custom iPXE as the "most generic do what you want" option and so we need to be able to handle it in Tinkebell.
+Custom Workflows (when supported by EM) should diminish the use of Custom iPXE, but I don't see it ever fully going away.
+Its also a great option to just boot into [netboot.xyz] to debug/explore some new hardware and so still useful for Tinkerbell itself.
+
+How to handle the 2 different types of OSes?
+Explicitly like in 1Left?
+Explicit and well defined is better than implicit or weakly defined, imo.
+Path 1Right just feels wrong, but is how we do things to day in EquinixMetal.
+
+Path 1Center is a bit of a mix, with benefits and warts from both sides.
+The common/well specified options are well specified, but leaves the generic bits out for interpretation.
+
+I lean towards trying out the 1Left path.
+
+#### Fork 2: To bitflags or not bitflags?
+
+Should we do bitflags in Instance.OperatingSystem.IP.flags (Path 2A) or break them out to bool `public` and `management` like currently done in [protos/packet][packet] (Path 2B)?
+
+I don't particularly like "exploding" flags into fields, but that could just be my once-upon-a-time embedded developer thinking leaking.
+
+I think another path/option is something like:
+```protobuf
+    enum Flags {
+      NONE = 0;
+      PUBLIC = 1;
+      MANAGEMENT = 2;
+    };
+    repeated Flags flags = 5;
+```
+Where the code will just accumulate the flags seen.
+
+This is a bit more formal of a definition of bitflags than "enums + int" but seems weird to me, I can see this being used though.
+This path doesn't protect against repeated dupes, but I doubt that would be a big deal.
+
+#### Extensions?
+
+I can already think of a couple of extensions that we'd need for EquinixMetal:
+* Licensing fields for some OSes
+* Spot Market related fields
+* IQN, used for block storage
+
+And the 2 ways I can think of to handle the extensions:
```

Comment:
I think it really depends on how the extensions are expected to be used/consumed.

If the goal is to only enable the ability to specify extra configuration for external (not a part of Tinkerbell) services, then I think the approaches mentioned below make sense. However, if we expect internal services to make use of the various extensions, then I think it might be better to encode them in through protobuf composition or make use of compatible types to provide a hacky approach to inheritance.